### PR TITLE
[Not for landing][Prototype] Add torch._record_scalar

### PR DESCRIPTION
**Pull request by yanboliang**
I'm adding this prototype to demonstrate a possible way that how intermediate logging works with torch.compile. The following example is a minimized repro and how it ladders up with proposed ```torch._record_scalar``` op. 
```
import torch
import torch.nn as nn
from typing import Tuple
from collections import deque
from functools import partial
from torch.utils.hooks import RemovableHandle

torch._dynamo.config.capture_scalar_outputs = True
torch._dynamo.config.capture_dynamic_output_shape_ops = True


my_config = dict()
my_config["MockModule"] = "mean"
my_config["MockModule.linear"] = "mean"
my_config["MockModule.relu"] = "mean"

record_path = "/tmp/record_scalar"


class MyLinear(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.weight = torch.nn.Parameter(torch.randn(out_features, in_features))
        self.bias = torch.nn.Parameter(torch.randn(out_features))

    def forward(self, x):
        return torch.nn.functional.linear(x, self.weight, self.bias)


class MockModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = MyLinear(10, 10)
        self.register_buffer("buf0", torch.randn(10, 10))

    def forward(self, x):
        return torch.nn.functional.relu(self.linear(x) + self.buf0)

def forward_hook(
    module: torch.nn.Module, inputs: torch.Tensor, output: torch.Tensor, prefix: str, aggregate_method: str
) -> torch.Tensor:
    if aggregate_method == "mean":
        torch._record_scalar(output.mean().item(), prefix, record_path)
    elif aggregate_method == "max":
        torch._record_scalar(output.max().item(), prefix, record_path)
    else:
        # demo purpose, using "min"
        torch._record_scalar(output.sum().item(), prefix, record_path)
    return output

def add_hooks(module, config):
    handles: List[RemovableHandle] = []
    q = deque([(module.__class__.__name__, module)])
    while q:
        name, m = q.pop()
        children = [(name + "." + n, y) for (n, y) in m.named_children()]
        q.extend(children)
        prefix = name + ":" + agg_method
        aggregate_method = config.get(name, "mean")
        handle = m.register_forward_hook(partial(forward_hook, prefix=prefix, aggregate_method=aggregate_method))
        if handle:
            handles.append(handle)
    return handles

x = torch.randn(10, 10, device="cuda")
mod = MockModule().to("cuda")

add_hooks(mod, my_config)

opt_mod = torch.compile(backend="inductor")(mod)
print(opt_mod(x))
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -183,6 +183,14 @@
...
     CompositeExplicitAutograd: _functional_assert_scalar
 
+- func: _record_scalar(Scalar self, str prefix, str filename) -> ()
+  dispatch:
+    CompositeExplicitAutograd: _record_scalar
+
+- func: _functional_record_scalar(Scalar self, str prefix, str filename, Tensor dep_token) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: _functional_record_scalar
+
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:193]: I don't think we need this form -- `_functional_assert_async.msg` was from a previous implementation of effect tokens that we don't use anymore

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -4680,6 +4681,44 @@ def codegen(self, wrapper):
             wrapper.writeline(f"{self.get_name()} = None")
 
 
+class RecordScalar(ExternKernel):
```
Comment (zou3519) [inline torch/_inductor/ir.py:4684]: Why can't this just go through the Inductor fallback mechanism? the custom op itself can, in its implementation, log the scalar to a file

Comment (yanboliang) [reply torch/_inductor/ir.py:4684]: Yes we can, but I'm doing this to flesh out ideas of further extension, if we allow ```_record_scalar``` to call python functions to log sth to a file, but probably it should be part of the callback op in the future.

Comment (zou3519) [reply torch/_inductor/ir.py:4684]: Inductor will probably have special handling for the with_effects HOP (cc @angelayi) (with_effects is basically just the callback op). So yeah that will likely get its own Kernel here

**Comment by yanboliang** [comment]
@ezyang Can I get your feedback about this idea?

**Comment by ezyang** [comment]
I'm not really sure what concretely is being proposed here. Let's talk about it at the offsite today?

**Pull request closed without merging**